crates/ruff_linter/src/rules/pylint/rules/dict_iter_missing_items.rs

```diff
@@ -0,0 +1,67 @@
+use ruff_python_ast::{Expr, ExprTuple};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_semantic::analyze::typing::is_dict;
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for unpacking a dictionary in a for loop without calling `.items()`.
+///
+/// ## Why is this bad?
+/// You are likely looking for an iteration over key, value pairs which can only be achieved
+/// when calling `.items()`.
+///
+/// ## Example
+/// ```python
+/// data = {"Paris": 2_165_423, "New York City": 8_804_190, "Tokyo": 13_988_129}
+/// for city, population in data:
+///     print(f"{city} has population {population}.")
+/// ```
+///
+/// Use instead:
+/// ```python
+/// data = {"Paris": 2_165_423, "New York City": 8_804_190, "Tokyo": 13_988_129}
+/// for city, population in data.items():
+///     print(f"{city} has population {population}.")
+/// ```
+#[violation]
+pub struct DictIterMissingItems;
+
+impl Violation for DictIterMissingItems {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Call `items()` when unpacking a dictionary for iteration")
+    }
+}
+
+pub(crate) fn dict_iter_missing_items(checker: &mut Checker, target: &Expr, iter: &Expr) {
+    let Expr::Tuple(ExprTuple { elts, .. }) = target else {
+        return;
+    };
+
+    if elts.len() != 2 {
+        return;
+    };
+
+    let Some(name) = iter.as_name_expr() else {
+        return;
+    };
+
+    let Some(binding) = checker
+        .semantic()
+        .only_binding(name)
+        .map(|id| checker.semantic().binding(id))
+    else {
+        return;
+    };
+    if !is_dict(binding, checker.semantic()) {
+        return;
+    }
+
+    checker
+        .diagnostics
+        .push(Diagnostic::new(DictIterMissingItems, iter.range()));
```

Comment:
We could also provide a fix if you're up to implement it :)

The fix could be to replace the identifier range with `{identifier}.items()`. You can either use the AST and generate the source code but I think it would be easier to just replace it with the string value like: `format!("{}.items()", identifier)`.

Reply:
I tried to keep the PR small as it is my first, but this proved to be really easy indeed. Added a fix. Classified it as safe as I think it is? But please let me know if it isn't.